### PR TITLE
Vueify Tool Form Data Selector

### PR DESCRIPTION
**Pull request by guerler**
This PR rewrites the data selector field for tool forms using Vue and TypeScript. It replaces the previous `select2` fields with `vue-multiselect`. The overall appearance of the input field has only been minimally adjusted, however there are some changes:

**Move File Browser to Button Group**
The file dialog button was previously placed to the right of the select field and is now shown in the button group on the left. The folder icon has been replaced with a dots/ellipsis to indicate that more datasets can be selected.

image

<br><br>

**Treat DCE's as separate source**
Dataset Collection Elements are not merged into the `hda` or `hdca` options list of the server response anymore but are instead treated as separate source. Server response in `basic.py` has been modified to account for that. However when it comes to displaying, the `dce`'s are like before in the corresponding hda/hdca select fields.

**TODO's**
- [x] Avoid flickering when switching between source types
- [x] Allow width resizing and word break in multiselect field
- [x] Augment dce handling to properly match batch modes
- [x] Add test cases, streamline source handling
- [x] Add handling for optional data fields if not already covered
- [x] Refine data matching when using file dialog

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by guerler** [comment]
Thanks for looking into it. I was wondering if this mechanism is also necessary for the ldda's. I have implemented it for now because that is how it was designed, but it might make more sense if the file dialog just imports the library datasets upfront and returns regular datasets and collection ids, instead of leaving that task to the tools api.

File: lib/galaxy/tools/parameters/basic.py
Comment (mvdbeek) [inline lib/galaxy/tools/parameters/basic.py]: Unless you're using this to determine if this represents a leaf in a collection this doesn't look right. If you do then this should just be a boolean (`is_dataset: true` ?).

**Review by mvdbeek — COMMENTED** [review]
> Dataset Collection Elements are not merged into or disguised as `hda` or `hdca` entries anymore but are instead treated as separate source.

This looks wrong on first glance, I think this would break establishing what kind of map over we're dealing with. Please don't merge this before I haven't reviewed this.

File: client/src/components/Form/Elements/FormData/FormData.vue
Comment (mvdbeek) [inline client/src/components/Form/Elements/FormData/FormData.vue]: what is element_id ? I think that was hacked in one of the backbone models, is that still used ?

**Comment by mvdbeek** [comment]
> `dce`'s are displayed when attempting to re-run a tool which consumed `dce`'s but since users cannot select other `dce`'s, the entry is displayed in a dismissible alert box instead of a select field. Dismissing the `dce` allows users to select an alternative input from the `hda` or `hdca` select fields.

Users will at least know if this was mapped a collection or not, and they can of course select another item (DCE or not) by dragging it in. I think this makes the rerun UX worse

File: lib/galaxy/tools/parameters/basic.py
Comment (guerler) [reply lib/galaxy/tools/parameters/basic.py]: Thanks for looking into this, and yes I am using this to determine if it's a dataset, and I think it's ok to keep it as `dce.hda`, feels more consistent since the same name is used in the backend.

**Comment by guerler** [comment]
> > `dce`'s are displayed when attempting to re-run a tool which consumed `dce`'s but since users cannot select other `dce`'s, the entry is displayed in a dismissible alert box instead of a select field. Dismissing the `dce` allows users to select an alternative input from the `hda` or `hdca` select fields.
> 
> Users used to at least know if this was mapped a collection, and they can of course select another item (DCE or not) by dragging it in. I think this makes the rerun UX worse

You can still drag another item into it.

File: lib/galaxy/tools/parameters/basic.py
Comment (guerler) [reply lib/galaxy/tools/parameters/basic.py]: I switched it over to a boolean named `is_dataset` as suggested.

File: client/src/components/Form/Elements/FormData/FormData.vue
Comment (guerler) [reply client/src/components/Form/Elements/FormData/FormData.vue]: Thanks, I removed the `element_id` check.

File: client/src/components/Form/FormDisplay.vue
Comment (ElectronicBlueberry) [inline client/src/components/Form/FormDisplay.vue]: Two things regarding this:
* use `structuredClone` instead of the JSON parse/stringify workaround to make a deep copy of an object
* I recently removed this, because it is a hacky way to force a reactive update from vue, which isn't always reliable. If you need something to update and it dosen't, that is because some data changed in a non-reactive way. Tracking this place down and replacing the data mutation with a `set` function triggers an update from vue without the need for a hack to force an update.

File: client/src/components/Form/FormRepeat.vue
Comment (ElectronicBlueberry) [inline client/src/components/Form/FormRepeat.vue]: This key will not track changes in array order.
`cacheId` is a bit of an misleading name, as this is simply an array index. Array indexes work fine for keying static lists, but break on dynamic ones. This was the reason for the object keying.

File: client/src/components/Form/FormDisplay.vue
Comment (guerler) [reply client/src/components/Form/FormDisplay.vue]: Yes, I noticed that this was deleted in https://github.com/galaxyproject/galaxy/pull/14892/files#diff-1cacf7869df763d8e0724de99083e02000e5964f59cefe3faca1d646192b8a1cL168 and although this PR is unrelated, I had to restore it in order to test this new input element. Unfortunately removing that line breaks the tool form e.g. #16602.

File: client/src/components/Form/FormRepeat.vue
Comment (guerler) [reply client/src/components/Form/FormRepeat.vue]: When restoring the reactivity of the tool form, using the `keyObject` tracking prevents the selection of datasets in repeats. However, I am ok with leaving this untouched in this PR since it is unrelated. I have removed the general fixes for the tool form from this PR. Thanks for looking into it.

File: client/src/components/Form/FormDisplay.vue
Comment (guerler) [reply client/src/components/Form/FormDisplay.vue]: I have removed that line, and restored the state of current dev. Thank you for looking into it.

File: client/src/components/Form/FormRepeat.vue
Comment (ElectronicBlueberry) [reply client/src/components/Form/FormRepeat.vue]: I think that bug should be easier to resolve once this PR is merged, without the need for the object deep copy

File: lib/galaxy_test/api/test_tools.py
Comment (mvdbeek) [inline lib/galaxy_test/api/test_tools.py]: As that's consumed by the API that's a breaking change, can you revert that please

**Review by mvdbeek — DISMISSED** [review]
We've discussed it on element as well, but to summarize it here:

Before:
Screenshot 2023-09-05 at 13 32 24

I see it's a single dataset, there's nothing special, it's all cool.

After:

image

Maybe you can infer it is a dataset, for whatever reason it's blue now and I have no clue what mode the job was run in.

This is a clear regression to me.

File: lib/galaxy_test/api/test_tools.py
Comment (guerler) [reply lib/galaxy_test/api/test_tools.py]: Renamed to `map_over_type` to avoid depreciation.

**Comment by guerler** [comment]
@mvdbeek thanks for the review and suggestions, I have reworked this. `dce`'s are still handled separately but displayed together with hda/hdca entries in the select fields.

**Comment by mvdbeek** [comment]
I don't think we should have `dce` in the interface. We used to just say `Select: <element_identifier>`. If you really want to knock it out of the park you could let users navigate to the parent collection and/or say `<element_identifier>: Element of Dataset Collection` or `<element_identifier>: Dataset Collection Element`

Screenshot 2023-09-09 at 15 32 19

**Comment by guerler** [comment]
@mvdbeek decided to replace `dce` with `as dataset` and `as dataset collection` depending on what the `dce` represents. I think we should leave more sophisticated labeling of `dce` to follow-ups.

File: client/src/components/Form/Elements/FormData/FormData.vue
Comment (mvdbeek) [inline client/src/components/Form/Elements/FormData/FormData.vue]: The use of an event bus here looks like a somewhat dubious shortcut, as does the somewhat buried passing of the props inside getValue. maybe it's fine though, it'd be good to have @galaxyproject/wg-uiux look at this.

**Review by mvdbeek — COMMENTED** [review]
Is the styling of the select content intentional ? It doesn't seem to align with our other colors
Screenshot 2023-09-15 at 14 34 20

It looks fine, but I don't feel qualified to make a judgement about the event bus. My gut feeling is that it's usually not a good idea, but maybe this is a good use ?

**Comment by guerler** [comment]
Yes, the styling is intentional, standard and entirely unrelated to this PR. This input element reuses the regular form select field. Styling options should be adjusted there. Regarding the `eventbus`, the usage of the `eventbus` avoids side effects and is cleaner than directly setting the value. It is also a non-critical feature since it is only used to show/hide the spinner. I think it is important to introduce such a critical input element change like this one early in the release cycle.

File: client/src/components/Form/Elements/FormData/FormData.vue
Comment (guerler) [reply client/src/components/Form/Elements/FormData/FormData.vue]: I removed the eventbus use case here, and instead pass a regular prop from the parent component to show/hide the spinner. I also removed `getValue` and instead directly inlined its code to avoid confusion.

**Comment by dannon** [comment]
Thanks for swapping out the eventBus and the other changes -- it's definitely better to use the standard interfaces/pattern here.  This is working pretty nicely in my testing; let's go ahead and get it into dev!

**Pull request merged**
